text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+[rpit]: https://rust-lang.github.io/impl-trait-initiative/explainer/rpit.html
+
+## Example use case
+
+The use case for `-> impl Trait` in trait functions is similar to its use in other contexts: traits often wish to return "some type" without specifying the exact type. As a simple example that we will use through the RFC, consider the `NewIntoIterator` trait, which is a variant of the existing `IntoIterator` that uses `impl Iterator` as the return type:
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+*This section assumes familiarity with the [basic semantics of impl trait in return position][rpit].*
+
+When you use `impl Trait` as the return type for a function within a trait definition or trait impl, the intent is the same: impls that implement this trait return "some type that implements `Trait`", and users of the trait can only rely on that.
+
+<!--However, the desugaring to achieve that effect looks somewhat different than other cases of impl trait in return position. This is because we cannot desugar to a type alias in the surrounding module; we need to desugar to an associated type (effectively, a type alias in the trait).-->
+
+Consider the following trait:
+
+```rust
+trait IntoIntIterator {
+    fn into_int_iter(self) -> impl Iterator<Item = u32>;
+}
+```
+
+The semantics of this are analogous to introducing a new associated type within the surrounding trait;
+
+```rust
+trait IntoIntIterator { // desugared
+    type IntoIntIter: Iterator<Item = u32>;
+    fn into_int_iter(self) -> Self::IntoIntIter;
+}
+```
+
+When using `-> impl Trait`, however, there is no associated type that users can name.
+
+By default, the impl for a trait like `IntoIntIterator` must also use `impl Trait` in return position.
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    fn into_int_iter(self) -> impl Iterator<Item = u32> {
+        self.into_iter()
+    }
+}
+```
+
+It can, however, give a more specific type with `#[refine]`:[^refine]
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    #[refine]
+    fn into_int_iter(self) -> impl Iterator<Item = u32> + ExactSizeIterator {
+        self.into_iter()
+    }
+    
+    // ..or even..
+
+    #[refine]
+    fn into_int_iter(self) -> std::vec::IntoIter<u32> {
+        self.into_iter()
+    }
+}
+```
+
+Users of this impl are then able to rely on the refined return type, as long as the compiler can prove this impl specifically is being used. Conversely, in this example, code that is generic over the trait can only rely on the fact that the return type implements `Iterator<Item = u32>`.
+
+### async fn desugaring
+
+`async fn` always desugars to a regular function returning `-> impl Future`. When used in a trait, the `async fn` syntax can be used interchangeably with the equivalent desugaring in the trait and trait impl:
```

Comment:
Presumably the reason for doing this is so that we can use `#[refine]` to make stricter promises on the impl? For example, adding a `Send` bound at the impl level.

Reply:
That's one (another example of an added promise would be that you don't capture `self`). A whole other reason would be to include code that runs at call time instead of during `poll`.